Update and fix big file upload (#4585)
owncloud/docs: modules/admin_manual/pages/configuration/files/big_file_upload_configuration.adoc

Please **do not** change `/usr/lib/systemd/system/httpd.service` directly, only use `sudo systemctl edit httpd`. If not doing so, a httpd package upgrade may revert your changes.
====

== Configuration via .htaccess/user.ini

NOTE: ownCloud comes with its own `owncloud/.htaccess` file. When using `php-fpm`, PHP settings in `.htaccess` are not accessed. These settings must then be set in the `{userini}` file. `php-fpm` will read settings from any {userini} file in the same directory as the .php file that is being served via a web server.

Set the following parameters inside the corresponding file using your own desired values, as in the following examples, both files are located in the ownCloud root folder:

{userini}::
[source,php]
----
post_max_size=16G
output_buffering=0
upload_max_filesize=16G
upload_tmp_dir=/mnt/php_big_temp/
----

{htaccess}::
[source,php]
----
php_value	post_max_size=16G
php_value	output_buffering=0
php_value	upload_max_filesize=16G
php_value	upload_tmp_dir=/mnt/php_big_temp/
----

> If you see PHP timeouts in your logfiles, increase the timeout values, which are in seconds, as in the example below, use the `php_value` prefix liek above when configuring the `{htaccess}` file:

Reviewer: ```suggestion
If you see PHP timeouts in your log files, increase the timeout values, which are in seconds, as in the example below. Use the `php_value` prefix like above when configuring the `{htaccess}` file:
```

[source,php]
----
max_input_time=3600
max_execution_time=3600
----

NOTE: Consider that any settings made in `{htaccess}` or `{userini}` may need to be repopulated after an upgrade of ownCloud.

== Configuring via PHP Global Settings

> If you don't want to use the ownCloud `.htaccess` or `.user.ini` file, you may configure PHP globally instead, if you have access to. 

Reviewer: ```suggestion
If you don't want to use the ownCloud `.htaccess` or `.user.ini` file, you may configure PHP globally instead. 
```

> Make sure to comment out or remove any lines in `.htaccess` added like in the section above, if you entered any.

Reviewer: ```suggestion
Make sure to comment out or remove any lines in `.htaccess` if you added any like in the section above.
```

NOTE: If you are running ownCloud on a 32-bit system, any `open_basedir` directive in your `php.ini` file needs to be commented out.

See the *Loaded Configuration File* section of xref:configuration/general_topics/general_troubleshooting.adoc#php-version-and-information[PHP Version and Information] to find your relevant php.ini files.

Set the following parameters inside the corresponding php.ini file using your own desired file size values, as in the following example:

[source,php]
----
post_max_size=16G
output_buffering=0
upload_max_filesize=16G
upload_tmp_dir=/mnt/php_big_temp/
----

> If you see PHP timeouts in your logfiles, increase the timeout values, which are in seconds, as in the example below:

Reviewer: ```suggestion
If you see PHP timeouts in your log files, increase the timeout values, which are in seconds, as in the example below:
```

[source,php]
----
max_input_time=3600
max_execution_time=3600
----

== Configuring via a Virtual Host

> You can configure php parameters also per virtual host - if you have access to the Apache configuration file. This eliminates the need maintaining custom settings in a `{userini}` or `{htaccess}` file especially on upgrades. Note the mandatory prefix `php_admin_value` before the php parameter.

Reviewer: ```suggestion
You can configure php parameters also per virtual host - if you have access to the Apache configuration file. This eliminates the need to maintain custom settings in a `{userini}` or `{htaccess}` file especially on upgrades. Note the mandatory prefix `php_admin_value` before the php parameter.
```

[source,apache]
----
<VirtualHost *:443>

	DocumentRoot /var/www/owncloud
	ServerName myowncloud.com

	php_admin_value	post_max_size 16G
	php_admin_value	output_buffering 0
	php_admin_value	upload_max_filesize 16G
	php_admin_value	upload_tmp_dir /mnt/php_big_temp/

	...
----

> If you see PHP timeouts in your logfiles, increase the timeout values, which are in seconds, as in the example below:

Reviewer: ```suggestion
If you see PHP timeouts in your log files, increase the timeout values, which are in seconds, as in the example below:
```

[source,php]
----
php_admin_value max_input_time 3600
php_admin_value max_execution_time 3600
----

== Configuring via ownCloud

As an alternative to the `upload_tmp_dir` of PHP (e.g., if you don't have access to your `php.ini`) you can also configure some parameters in `config.php`.

* Set a temporary location for uploaded files by using the `tempdirectory` setting.
* If you have configured the `session_lifetime` setting in your `config.php`, 
see xref:configuration/server/config_sample_php_parameters.adoc[Sample Config PHP Parameters], make sure it is not too low. This setting needs to be configured to at least the time (in seconds) that the longest upload will take. If unsure, remove this entirely from your configuration to reset it to the default shown in the `config.sample.php`.

== General Upload Issues

Various environmental factors could cause a restriction of the upload size. Examples are:

* The `LVE Manager` of `CloudLinux` which sets an `I/O limit`.
* Some services like `Cloudflare` are also known to cause uploading issues.
* Upload limits enforced by proxies used by your clients.

> * Other webserver modules like described in xref:configuration/general_topics/general_troubleshooting.adoc[General Troubleshooting].

Reviewer: ```suggestion
* Other web server modules like described in xref:configuration/general_topics/general_troubleshooting.adoc[General Troubleshooting].
```

=== Apache Directives

* {limitrequestbody-url}[LimitRequestBody]
* {sslrenegbuffersize-url}[SSLRenegBufferSize]

=== Apache with mod_reqtimeout

> The {mod_reqtimeout-url}[mod_reqtimeout] Apache module could also stop large uploads from completing. If you're using this module and getting failed uploads of large files, either disable it in your Apache config or raise the configured `RequestReadTimeout` timeouts.

Reviewer: ```suggestion
The {mod_reqtimeout-url}[mod_reqtimeout] Apache module could also stop large uploads from completing. If you're using this module and uploads of large files fail, either disable it in your Apache config or increase the configured `RequestReadTimeout` values.
```

> Disable mod_reqtimeout On Ubuntu::

Reviewer: ```suggestion
Disable mod_reqtimeout on Ubuntu::
```

+
On Ubuntu, you can disable the module by running the following command:
+
[source,console]
----
sudo a2dismod reqtimeout
----

> Disable mod_reqtimeout On CentOS::

Reviewer: ```suggestion
Disable mod_reqtimeout on CentOS::
```

+
On CentOS, comment out the following line in `/etc/httpd/conf/httpd.conf`:
+
[source,apache]
----
LoadModule reqtimeout_module modules/mod_reqtimeout.so
----
+

> When you have done run `asdismod` or updated `/etc/httpd/conf/httpd.conf`, restart Apache.

Reviewer: ```suggestion
When you have run `asdismod` or updated `/etc/httpd/conf/httpd.conf`, restart Apache.
```

+

> TIP: There are also several other configuration options in your web server config which could prevent the upload of larger files. Please see your web server's manual, for how to configure those values correctly:

Reviewer: ```suggestion
TIP: There are also several other configuration options in your web server config which could prevent the upload of larger files. Refer to your web server's manual for how to configure those values correctly:
```

=== Apache with mod_fcgid

* {fcgidmaxrequestinmem-url}[FcgidMaxRequestInMem]
* {fcgidmaxrequestlen-url}[FcgidMaxRequestLen]

> WARNING: If you are using Apache/2.4 with mod_fcgid, as of February/March 2016, `FcgidMaxRequestInMem` still needs to be significantly increased from its default value to avoid the occurrence of segmentation faults when uploading big files. This is not a regular setting but serves as a workaround for {mod_fcgid_bug_51747-url}[Apache with mod_fcgid bug #51747].

Reviewer: ```suggestion
WARNING: If you are using Apache 2.4 with mod_fcgid, as of February/March 2016, `FcgidMaxRequestInMem` still needs to be significantly increased from its default value to avoid the occurrence of segmentation faults when uploading big files. This is not a regular setting but serves as a workaround for {mod_fcgid_bug_51747-url}[Apache with mod_fcgid bug #51747].
```
